### ISSUE TITLE
cBioportal requires to fix a critical error in authentication disregarding user's password

### ISSUE THREAD

**Issue by gatksun** [cBioPortal/cbioportal#7496]
I set up a local cBioportal instance since its 1.x. 
Initially, my cBioportal applied ad as its authentication.  Months ago, I read your posting that mentions how to use ldap to authenticate, by the method, so  I changed my authentication to ldap.
However, recently I found the authentication with deadly errors no matter ad or ldap. I think this error existing in all authentications. Of course, I can't test other authentication than ad and ldap.
When an user log in with username, he/she can use any word as password to enter cBioportal.
I traced the log files. I found this error is caused by security/security-spring/src/main/resources/applicationContext-security.xml
Each authentication includes the below setting:
<authentication-manager alias="authenticationManager">
        
        <authentication-provider ref="tokenUserDetailsAuthenticationProvider"/>
        <authentication-provider ref="xxxxAuthenticationProvider"/>
 </authentication-manager>

Since tokenUserDetailsAuthenticationProvider exists, any authentication-provider following it can't override it completely. 
When any authentication as ad or ldap is going verification, it got a correct username, but password or other reason failed, then authentication instead would use "tokenUserDetailsAuthenticationProvider" to continue successful.
Thus, as long as users know correct username, no matter what password is used, log in is okay.
If I remove <authentication-provider ref="tokenUserDetailsAuthenticationProvider"/>, the password is required, that is correct authentication!
On your posting or google group regarding ldap's authentication, it is wrong configuration, at least no work by what I tested.
I do not know from which version, you start to add <authentication-provider ref="tokenUserDetailsAuthenticationProvider"/>, wrong authentication just started this configuration. Look forward your fixing it soon.
Since your cbioportal is public without authentication, users do not care of this issue.
Another request: can you add the below two columns in data table users?
logdate : datetime, when an user logs in, update it to current time.
logged: int, count logged time for an user.

Harry Sun

**Comment by kalletlak** [comment]
@gatksun we don't have setup for ad/ldap. Could you try adding  `authentication-failure-url="/login.jsp?login_error=true"` to `form-login` at https://github.com/cBioPortal/cbioportal/blob/dc7a335bf2d78bd0b2057ae69b7f8472f1e78cfe/security/security-spring/src/main/resources/applicationContext-security.xml#L529 and https://github.com/cBioPortal/cbioportal/blob/dc7a335bf2d78bd0b2057ae69b7f8472f1e78cfe/security/security-spring/src/main/resources/applicationContext-security.xml#L564 and see if it works

**Comment by gatksun** [comment]
No work after I added authentication-failure-url="/login.jsp?login_error=true"
I do not think this bug is only related to ad,ldap, should do with all authenticatons.
I found authentication-failure-url="/login.jsp?login_error=true" added to openid.
If you used any authentication, you may log in with correct username and wrong password. Note that openid is a little different.
It works when I still remove authentication-provider ref="tokenUserDetailsAuthenticationProvider"
Thanks
     On Friday, May 15, 2020, 07:58:19 AM CDT, Karthik Kalletla <notifications@github.com> wrote:  
 
 


@gatksun we don't have setup for ad/ldap. Could you try adding authentication-failure-url="/login.jsp?login_error=true" to form-login at https://github.com/cBioPortal/cbioportal/blob/dc7a335bf2d78bd0b2057ae69b7f8472f1e78cfe/security/security-spring/src/main/resources/applicationContext-security.xml#L529 and https://github.com/cBioPortal/cbioportal/blob/dc7a335bf2d78bd0b2057ae69b7f8472f1e78cfe/security/security-spring/src/main/resources/applicationContext-security.xml#L564 and see if it works

**Comment by kalletlak** [comment]
Thanks for checking out. This issue might be just with ad and ldap authentications. Other authentications methods are working fine.

**Comment by gatksun** [comment]
You should try easily ldap if your mskcc uses active directory supporting ldap, just set the below ldap:

## configuration for the LDAP access
ldap.user_search_base=dc=mskcc,dc=org
ldap.url=ldap://ldap.mskcc.org:389
ldap.manager.dn=
ldap.manager.password=
## The following attributes for are good for ActiveDirectory, for OpenLDAP use "uid" for username
ldap.attributes.username=sAMAccountName
#ldap.attributes.username=uid
ldap.attributes.last_name=sn
ldap.attributes.given_name=givenName
ldap.attributes.email=mail


The above ldap works , but wrong because it accepts wrong password.
Actually, ldap in cbioportal may be totally wrong. If I removed authentication-provider ref="tokenUserDetailsAuthenticationProvider", it no longer works even by correct username and password.     On Friday, May 15, 2020, 10:55:39 AM CDT, Karthik Kalletla <notifications@github.com> wrote:  
 
 


Thanks for checking out. This issue might be just with ad and ldap authentications. Other authentications methods are working fine.

**Comment by kalletlak** [comment]
@gatksun I just setup ldap locally and tested, everything seems to be working fine. I used  [OpenDJ](https://github.com/OpenIdentityPlatform/OpenDJ) to set it ldap locally.

CC: @jjgao @n1zea144

**Comment by gatksun** [comment]
After I logged in v3.3.3 by ad authentication, its Logged in as xxxx disappeared initially. If I reloaded the page, it shows up.
If you used ldap as authentication, can you log in with correct username and wrong password?
     On Wednesday, May 20, 2020, 09:57:03 AM CDT, Karthik Kalletla <notifications@github.com> wrote:  
 
 


@gatksun I just setup ldap locally and tested, everything seems to be working fine. I used OpenDJ to set it ldap locally.

CC: @jjgao @n1zea144

**Comment by gatksun** [comment]
I do not believe you fixed it if you didn't change any codes. Why did you close it? If you use ldap as authentication, please try to log in with correct username and wrong password. If you can't log in, please send your applicationContext-security.xml and portal.properties to me. Thank you. I posted this issue, just hope you can fix it.
     On Wednesday, May 20, 2020, 09:57:02 AM CDT, Karthik Kalletla <notifications@github.com> wrote:  
 
 


Closed #7496.

**Comment by kalletlak** [comment]
There are no code changes. I tested on latest release. I've tried all combinations and everything works fine.
I used ldap authentication method
Here are the configurations what I set

ldap.user_search_base=dc=example,dc=com
ldap.url=ldap://test.home:389
ldap.manager.dn=cn=Directory Manager
ldap.manager.password=******
ldap.attributes.username=uid
ldap.attributes.last_name=sn
ldap.attributes.given_name=givenName
ldap.attributes.email=mail

**Comment by gatksun** [comment]
I tested ldap through 3.3.3 based on our active directory server. Actually, no required for Manager and password, also no required for user's correct password, because it may retrace to tokenUserDetailsAuthenticationProvider. For my local instance, I already fixed this issue I told you before. I do think this is a bug. If do not think so, keep it.
     On Wednesday, May 20, 2020, 11:20:24 AM CDT, Karthik Kalletla <notifications@github.com> wrote:  
 
 


There are no code changes. I tested on latest release. I've tried all combinations and everything works fine.
I used ldap authentication method
Here are the configurations what I set

ldap.user_search_base=dc=example,dc=com
ldap.url=ldap://test.home:389
ldap.manager.dn=cn=Directory Manager
ldap.manager.password=******
ldap.attributes.username=uid
ldap.attributes.last_name=sn
ldap.attributes.given_name=givenName
ldap.attributes.email=mail

**Comment by kalletlak** [comment]
@gatksun seems to be working fine even without setting ldap manager dn and password. It would stay on the login page when wrong password is entered.

Thanks
Karthik

**Comment by gatksun** [comment]
Yes, why does it accept wrong password? because <authentication-provider ref="tokenUserDetailsAuthenticationProvider"/> is set before the other authentication provider. Even though real authentication provider failed, the authentication will use tokenUserDetailsAuthenticationProvider to go through everything to cbioportal. 
Simply, as long as you remove <authentication-provider ref="tokenUserDetailsAuthenticationProvider"/> in all authentications, you can't use wrong password or wrong authentication to enter the cbioportal.
Actually, ldap provided in cbioportal does not work !
 
     On Wednesday, May 20, 2020, 04:00:59 PM CDT, Karthik Kalletla <notifications@github.com> wrote:  
 
 


@gatksun seems to be working fine even without setting ldap manager dn and password. It would stay on the login page when wrong password is entered.

Thanks
Karthik
